### PR TITLE
Add spaces visitor

### PR DESCRIPTION
**Pull request by kunli2**
Introduce a space visitor to support all Kotlin specific space code styles.

### CODE REVIEW

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaSourceFile;
+import org.openrewrite.kotlin.KotlinIsoVisitor;
+import org.openrewrite.kotlin.style.IntelliJ;
+import org.openrewrite.kotlin.style.SpacesStyle;
+
+import static java.util.Objects.requireNonNull;
+
+public class Spaces extends Recipe {
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/Spaces.java:28]: We don't need a recipe just for spaces, we need a full `AutoFormat`. This should be the start of the kotlin `AutoFormat`.

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        public @Nullable J visit(@Nullable Tree tree, ExecutionContext executionContext) {
+            if (tree instanceof JavaSourceFile) {
+                JavaSourceFile cu = (JavaSourceFile) requireNonNull(tree);
+                SpacesStyle style = ((SourceFile) cu).getStyle(SpacesStyle.class);
+                if (style == null) {
+                    style = IntelliJ.spaces();
+                }
+                doAfterVisit(new SpacesVisitor<>(style));
+            }
+            return super.visit(tree, executionContext);
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/Spaces.java:56]: Wasteful to continue to visit the sub-elements of the current tree

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new SpacesFromCompilationUnitStyle();
+    }
+
+    private static class SpacesFromCompilationUnitStyle extends KotlinIsoVisitor<ExecutionContext> {
+        @Override
+        public @Nullable J visit(@Nullable Tree tree, ExecutionContext executionContext) {
+            if (tree instanceof JavaSourceFile) {
+                JavaSourceFile cu = (JavaSourceFile) requireNonNull(tree);
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/Spaces.java:49]: `instanceof` check guarantees non-null already, redundant to also call `requireNonNull`

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 2 above)
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/Spaces.java:56]:
```suggestion
            if (!(tree instanceof K.CompilationUnit)) {
                return tree;
            }
            K.CompilationUnit cu = (K.CompilationUnit) tree;
            SpacesStyle style = cu.getStyle(SpacesStyle.class, IntelliJ.spaces());
            return new SpacesVisitor<>(style).visitNonNull(cu, getCursor().fork());
```

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            return super.visit(tree, executionContext);
+        }
+    }
+
+    public static <J2 extends J> J2 formatSpaces(J j, Cursor cursor) {
+        SourceFile cu = cursor.firstEnclosingOrThrow(SourceFile.class);
+        SpacesStyle style = cu.getStyle(SpacesStyle.class);
+        //noinspection unchecked
+        return (J2) new SpacesVisitor<>(style == null ? IntelliJ.spaces() : style).visitNonNull(j, 0, cursor);
+    }
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/Spaces.java:65]: We don't need this method.
Instead `KotlinVisitor` should override `autoformat()` and `maybeAutoformat()` methods from `JavaVisitor` to invoke kotlin auto format

File: src/main/java/org/openrewrite/kotlin/format/SpacesVisitor.java [hunk 5]
```diff
@@ -0,0 +1,1232 @@
...
+
+    @Override
+    public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, P p) {
+        J.ClassDeclaration c = super.visitClassDeclaration(classDecl, p);
+        if (c.getBody() != null) {
+            c = c.withBody(spaceBefore(c.getBody(), beforeLeftBrace));
+            if (c.getBody().getStatements().isEmpty()) {
+                if (c.getKind() != J.ClassDeclaration.Kind.Type.Enum) {
+                    // Defaulted to `false` in IntelliJ's Kotlin formatting.
+                    boolean withinCodeBraces = false;
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/format/SpacesVisitor.java:211]: These should be hardcoded, not read from the style

File: src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java [hunk 6]
```diff
@@ -858,10 +858,6 @@ public J visitTernary(J.Ternary ternary, PrintOutputCapture<P> p) {
         @Override
         public J visitTypeCast(J.TypeCast typeCast, PrintOutputCapture<P> p) {
             beforeSyntax(typeCast, Space.Location.TYPE_CAST_PREFIX, p);
-            boolean printParens = !typeCast.getMarkers().findFirst(OmitParentheses.class).isPresent();
-            if (printParens) {
-                p.append('(');
-            }
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java:864]: How does this not break kotlin code that uses lambdas?
```kotlin
listOf(1,2,3).forEach { println(it) } // very common to omit parentheses here
```

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 2 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java:56]: I see, it should be
```java 
return (J) tree; 
```

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 1 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java:28]: I think this file is copied from `rewrite`, I checked the code in the rewrite, and it seems this is only used for the convenience of test purposes. I feel we should remove `Spaces` recipe but keep the internal static class `SpacesFromCompilationUnitStyle` which includes style detector.

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 3 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java:49]: revised.

File: src/main/java/org/openrewrite/kotlin/format/SpacesVisitor.java (hunk 5 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/SpacesVisitor.java:211]: Sure, I am going to clean up these kind of hard-coded stuffs after moving all space tests to `rewrite-kotlin`

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 4 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java:62]: yes, this is copied from rewrite and not used anywhere, removed.

File: src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java (hunk 6 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java:864]: I am not sure I understand this question, 
I think for lambda, it's handled by `visitLambda` method in this file.

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 1 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java]: removed this file, and have `SpacesFromCompilationUnitStyle` instead.

File: src/main/java/org/openrewrite/kotlin/format/Spaces.java (hunk 4 above)
Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/format/Spaces.java]: We haven't had an `AutoFormatVisitor` yet but just introduced SpacesVisitor only in this PR. 
will make the change of override `autoformat()` and `maybeAutoformat()` methods in `KotlinVisitor` in a separate PR

File: src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java (hunk 6 above)
Comment (sambsnyd) [reply src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java:864]: Oh yeah this is type casts.
In kotlin there are two kinds of type casts, the java-style `(Foo) bar` and `bar as Foo`. Does the latter still print correctly with this change?

Comment (kunli2) [reply src/main/java/org/openrewrite/kotlin/internal/KotlinPrinter.java:864]: Yeah, this code is actually removed from @traceyyoshima  initial change, he might have more context about why remove this, but I verified the printer works for the latter for the below case. 
```kotlin
fun main() {
    val obj: Any = "Hello"
    val str: String = obj as String
}
```
but I am not sure how to do java-style type case in Kotlin.
I tried this but doesn't work
```kotlin
fun main() {
    val obj: Any = "Hello"
    // val str: String = obj as String
    val str2: String = (String) obj
}
```

**Pull request merged**